fetch/fetch.1
.Dd May 15, 2020
.Dt HSCRIPT-FETCH 1
.Os "Adélie Linux"
.Sh NAME
.Nm hscript-fetch
.Nd fetch a HorizonScript for further processing
.Sh SYNOPSIS
.Nm
.Ar FILE-OR-URL
.Sh DESCRIPTION
The
.Nm
utility retrieves the HorizonScript from the location specified by
.Ar FILE-OR-URL ,
and writes it to the path expected by the
.Xr hscript-executor 8
utility; presently,
.Pa /etc/horizon/installfile .
.Pp
The single argument,
.Ar FILE-OR-URL ,
should specify either an absolute path (one that begins with a /), or a URL.
.Ss Remote script retrieval
If a remote URL is specified, the
.Nm
utility will attempt to download the HorizonScript from it.  Supported URL
protocols include HTTP, HTTPS, and TFTP.
.Sh FILES
.Bl -ohang -width "/etc/horizon/installfile" -offset indent -compact
.It Pa /etc/horizon/installfile
.Nm
writes the HorizonScript specified to
.Pa /etc/horizon/installfile .
.Sh EXIT STATUS
.Ex -std
.Sh EXAMPLES
hscript-fetch /media/live/desktop.installfile
.Pp
hscript-fetch https://deploy.company.int/workstation16.installfile
.Sh DIAGNOSTICS
.Bl -diag
.It "%dateT%time log %location: %status: %message[: %extra]"
A message of type
.Cm %status
was logged while processing
.Cm %location .
The log message
.Cm %message
may be followed by additional information or hints in
.Cm %extra .
.El
.Sh SEE ALSO
.Xr hscript 5 .
.Sh HISTORY
The
.Nm
command first appeared in Horizon 0.1.
.Sh AUTHORS
.An A. Wilcox
.Aq awilfox@adelielinux.org
.Sh SECURITY CONSIDERATIONS
The
.Nm
utility supports unencrypted HTTP download.  Extreme caution should be
exercised when using unencrypted HTTP downloads, as your system's root
password could be manipulated by a malicious attacker.  Please use HTTPS
whenever possible.


CHANGELOG.rst
===============================
 Changelog for Project Horizon
===============================
:Author:
  * **A. Wilcox**, documentation writer
  * **Contributors**, code
:Copyright:
  © 2019-2020 Adélie Linux and contributors.


0.9.0 (2020-05-xx)
==================

Build
-----

* The BCNM check is now functional.

Disk
----

* HFS+ has been added as a supported file system.

* disklabel: MBR support has been fixed.

* partition: Add 'bios' and 'prep' flags.

* partition: Type codes are now handled properly.

* partition: Ensure the first partition isn't too close to the first sectors.

* fs: Undo files for e2fs are no longer created.

* mount: Each mount key no longer overwrites /etc/fstab in the target.

DiskMan
-------

* A new library for probing disk information, DiskMan, has been written.

Documentation
-------------

* All tools and libraries, except libhscript, have been documented with
  manual pages.

HorizonScript Library
---------------------

* Introspection support has been added to the Script class.

* Refactored Keys to be owned by a Script, so Keys can access the values of
  other keys.

* The target directory is now configurable.

* /dev, /proc, and /sys are now mounted in the target.

Image Creation
--------------

* A new system for creating images using the Horizon system has been written.

Metadata
--------

* arch: New key added, including specification and implementation.

* repository: Fixed defaults when 'firmware' is set to true.

* timezone: Fixed issue when target already contained /etc/localtime.

Network
-------

* Configure network interfaces inside the Install Environment when
  'network' is set to true.

* netconfigtype: New key added, including specification and implementation.

* Existing network configuration on the Installation Environment system is
  now properly handled.

Package
-------

* APK Tools are now invoked more efficiently.

* --keys-dir is now passed to APK Tools during base installation.

Project
-------

* A Code of Conduct has been added.

* A YANG model, describing a JSON schema for representing a HorizonScript,
  has been written.

Tests
-----

* A test has been added for IPv4 subnet -> CIDR conversion.

Tools
-----

* Use Boost's program_options instead of vendoring clipp.  Thanks to
  Calvin Buckley.

* Messages have been unified between the CLI tools, to ensure consistency.

* A new tool, hscript-fromjson, has been added to convert JSON representations
  of HorizonScripts to HorizonScript.

UI
--

* The Qt 5 UI has been added.  It is not yet fully implemented, but basic
  installations should work.

User
----

* userpassphrase: Ensure simulated runs don't actually set passwords.

* All user keys now run shadow commands in the target, instead of the
  Installation Environment system.

Util
----

* Factor subnet -> CIDR conversion to util function.




0.2.0 (2019-11-07)
==================

Disk
----

* lvm_pv, lvm_vg, and lvm_lv execution are now implemented.


Metadata
--------

* keymap execution is now implemented.

* language: An issue with execution of the language key has been fixed.

* signingkey: Firmware keys are now installed when firmware is true.


Network
-------

* hostname: dns_domain_lo is now properly set in target /etc/conf.d/net.

* nameserver execution is now implemented.

* netaddress: OpenRC services are now added for configured interfaces.


Owner
-----

* New utility 'hscript-printowner' added, which prints the owning UID of a
  given path.


User
----

* User account creation is now fully implemented.




0.1.0 (2019-11-02)
==================

Initial release.


fetch/fetch.cc
/*
 * fetch.cc - Implementation of the HorizonScript Locator
 * Project Horizon
 *
 * Copyright (c) 2019 Adélie Linux and contributors.  All rights reserved.
 * This code is licensed under the AGPL 3.0 license, as noted in the
 * LICENSE-code file in the root directory of this repository.
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

#include <algorithm>        /* transform */
#include <cstdlib>          /* EXIT_* */
#ifdef HAVE_LIBCURL
#    include <cstring>      /* strerror */
#    include <curl/curl.h>  /* curl_* */
#    include <errno.h>      /* errno */
#endif /* HAVE_LIBCURL */
#include <fstream>
#include <map>
#include <string>
#include <unistd.h>         /* access */
#include "util/output.hh"

static const char *IFILE_PATH = "/etc/horizon/installfile";
static const int SCRIPT_LINE_MAX = 512;

bool pretty = true;

/*! Process a local path and copy it to the proper location.
 * @param path      The local path.
 * @returns An exit code.
 */
int process_local(const std::string &path) {
    /* if we can't read it, don't even bother */
    if(access(path.c_str(), R_OK) != 0) {
        output_error("process_local",
                     std::string(path) + " does not exist or is not readable");
        return EXIT_FAILURE;
    }

    std::ofstream output(IFILE_PATH, std::ios_base::trunc);
    if(!output) {
        output_error("process_local",
                     std::string(IFILE_PATH) +
                     " could not be opened for writing");
        return EXIT_FAILURE;
    }
    std::ifstream input(path, std::ios_base::in);
    if(!input) {
        output_error("process_local",
                     path + " could not be opened for reading");
        return EXIT_FAILURE;
    }
    char buffer[SCRIPT_LINE_MAX];

    while(input.getline(buffer, SCRIPT_LINE_MAX)) {
        output << buffer << std::endl;
    }

    if(input.fail() && !input.eof()) {
        output_error("process_local", "line too long while reading " + path);
        return EXIT_FAILURE;
    }
    if(input.bad() && !input.eof()) {
        output_error("process_local", "I/O error reading " + path);
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}


#ifdef HAVE_LIBCURL
/*! Download an installfile using cURL.
 * @param path      The remote path to download.
 */
int process_curl(const std::string &path) {
    CURL *curl = curl_easy_init();
    CURLcode result;
    int return_code = EXIT_FAILURE;
    char errbuf[CURL_ERROR_SIZE];
    FILE *fp;

    if(curl == nullptr) {
        output_error("internal", "trouble initialising cURL library");
        return EXIT_FAILURE;
    }

    fp = fopen(IFILE_PATH, "w");
    if(fp == nullptr) {
        output_error("internal", "couldn't open " + std::string(IFILE_PATH) +
                     " for writing", strerror(errno));
        curl_easy_cleanup(curl);
        return EXIT_FAILURE;
    }

    curl_easy_setopt(curl, CURLOPT_URL, path.c_str());
    curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, errbuf);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);

    result = curl_easy_perform(curl);
    if(result == CURLE_OK) {
        return_code = EXIT_SUCCESS;
    } else {
        output_error("curl", "couldn't download installfile", errbuf);
    }
    curl_easy_cleanup(curl);
    return return_code;
}
#endif /* HAVE_LIBCURL */


/*! The signature of a protocol dispatch function */
typedef int (*proto_dispatch_fn)(const std::string &);


/*! Protocol handlers */
static const std::map<std::string, proto_dispatch_fn> protos = {
#ifdef HAVE_LIBCURL
    {"http", process_curl},
    {"https", process_curl},
    {"tftp", process_curl},
#endif /* HAVE_LIBCURL */
};


/*! Dispatch handling of +path+ to a protocol handler, or return failure.
 * @param path      The remote path to download to IFILE_PATH.
 * @returns EXIT_SUCCESS if the file was downloaded; EXIT_FAILURE otherwise.
 * @note Currently, it is ambiguous whether EXIT_FAILURE means invalid proto
 * or error downloading from the URL.
 */
int process_maybe_remote(const std::string &path) {
    std::string proto = path.substr(0, path.find_first_of(":"));
    std::transform(proto.begin(), proto.end(), proto.begin(), ::tolower);

    if(protos.find(proto) != protos.end()) {
        return (*protos.at(proto))(path);
    }

    std::string support = "This build of Horizon supports: ";
    for(auto &proto_pair : protos) {
        support += proto_pair.first + " ";
    }

    output_error("argv", proto + " is not a valid protocol", support);
    return EXIT_FAILURE;
}


/*! Figure out the TFTP path(s) to try, and use them */
int full_auto_tftp() {
    output_error("tftp", "stub!");
    return EXIT_FAILURE;
}


/*
 * The goal of the Locator is to find the HorizonScript for this computer,
 * which is the target, and copy it to the well-known path /etc/horizon
 * as 'installfile'.
 */
int main(int argc, char *argv[]) {
    std::string installfile;

    if(argc < 1 || argc > 2) {
        std::cerr << "usage: " << std::endl;
        std::cerr << "\thscript-fetch [path|url]" << std::endl;
        return EXIT_FAILURE;
    }

    bold_if_pretty(std::cout);
    std::cout << "HorizonScript Locator version " << VERSTR;
    reset_if_pretty(std::cout);
    std::cout << std::endl
              << "Copyright (c) 2019-2020 Adélie Linux and contributors."
              << std::endl
              << "This software is licensed to you under the terms of the "
              << std::endl << "AGPL 3.0 license, unless otherwise noted."
              << std::endl << std::endl;

    if(argc == 1) {
        if(access(IFILE_PATH, F_OK) == 0) {
            /* That was easy™ */
            std::cout << "HorizonScript already present at " << IFILE_PATH
                      << ".  Quitting." << std::endl;
            return EXIT_SUCCESS;
        }

        if(protos.find("tftp") == protos.end()) {
            output_error("internal", "A Fully Automatic installation requires "
                         "Horizon to be built with TFTP support");
            return EXIT_FAILURE;
        } else {
            return full_auto_tftp();
        }
    }

    std::string path(argv[1]);
    if(path == "--version") {
        /* Banner printed above is our version statement. */
        return EXIT_SUCCESS;
    }

    if(path[0] == '/') {
        return process_local(path);
    }
    if(path.find("://") != std::string::npos) {
        return process_maybe_remote(path);
    }

    output_error("argv", "Unrecognised protocol or invalid URL",
                 "An absolute path or URL is required");
    return EXIT_FAILURE;
}
